ddtrace/ext/provider.py

```diff
@@ -0,0 +1,75 @@
+"""
+common CI providers
+"""
+
+import os
+import re
+
+from ddtrace.vendor import attr
+
+from . import ci, git
+from ..internal.logger import get_logger
+
+log = get_logger(__name__)
+
+_PROVIDER_METADATA_KEY = "ddtrace.ext.provider"
+
+
+def field(name, *args, **kwargs):
+    kwargs.setdefault("metadata", {})
+    kwargs["metadata"][_PROVIDER_METADATA_KEY] = name
+    return attr.ib(*args, **kwargs)
+
+
+_RE_BRANCH_PREFIX = re.compile(r"^refs/(heads/)?")
+
+
+@attr.s(kw_only=True, eq=False, order=False, slots=True)
+class Provider(object):
+
+    # CI properties
+    job_url = field(ci.JOB_URL, default=None)
+    pipeline_id = field(ci.PIPELINE_ID, default=None)
+    pipeline_name = field(ci.PIPELINE_NAME, default=None)
+    pipeline_number = field(ci.PIPELINE_NUMBER, default=None)
+    pipeline_url = field(ci.PIPELINE_URL, default=None)
+    provider_name = field(ci.PROVIDER_NAME, default=None)
+    workspace_path = field(ci.WORKSPACE_PATH, default=None)
+
+    # Git properties
+    branch = field(
+        git.BRANCH, default=None, converter=attr.converters.optional(lambda value: _RE_BRANCH_PREFIX.sub("", value))
+    )
+    commit_sha = field(git.COMMIT_SHA, default=None)
+    repository_url = field(git.REPOSITORY_URL, default=None)
+    tag = field(git.TAG, default=None)
+
+    # Deprecated properties
+    _deprecated_commit_sha = field("git.commit_sha")
```

Comment:
@Kyle-Verhoog this is the reason why I would like to keep this level of complexity.